### ISSUE TITLE
Cannot build with nrf51-hal

### ISSUE THREAD

**Issue by nedv-eu** [knurling-rs/app-template#59]
When I try to build for nrf51822 (cortex-m0) with nrf-51-hal crate the build never ends.

Steps to reproduce:

rustup target add thumbv6m-none-eabi
cargo generate  --git https://github.com/knurling-rs/app-template  --branch main   --name my-app
Add nrf51-hal = "0.14" to Cargo.toml dependencies
cargo build

cargo build output:
   Compiling proc-macro2 v1.0.34
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.82
   Compiling semver-parser v0.7.0
   Compiling cortex-m v0.7.3
   Compiling version_check v0.9.3
   Compiling nb v1.0.0
   Compiling vcell v0.1.3
   Compiling void v1.0.2
   Compiling cortex-m-rt v0.7.1
   Compiling bitfield v0.13.2
   Compiling defmt v0.3.0
   Compiling az v1.2.0
   Compiling typenum v1.14.0
   Compiling defmt-macros v0.3.1
   Compiling fixed v1.11.0
   Compiling nrf51-pac v0.10.1
   Compiling defmt-parser v0.3.0
   Compiling half v1.8.2
   Compiling stable_deref_trait v1.2.0
   Compiling bitflags v1.3.2
   Compiling nrf-hal-common v0.14.0
   Compiling bytemuck v1.7.3
   Compiling cast v0.3.0
   Compiling rand_core v0.6.3
   Compiling cfg-if v1.0.0
   Compiling defmt-rtt v0.3.1
   Compiling panic-probe v0.3.0
   Compiling nrf51-hal v0.14.0
   Compiling embedded-storage v0.2.0
   Compiling volatile-register v0.2.1
   Compiling nb v0.1.3
   Compiling semver v0.9.0
   Compiling embedded-dma v0.1.2
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling embedded-hal v0.2.6
   Compiling rustc_version v0.2.3
   Compiling bare-metal v0.2.5
   Compiling quote v1.0.10
   Compiling cortex-m-rt-macros v0.7.0
   Compiling my-app v0.1.0 (/home/honza/tmp/my-app)
    Building [=======================>   ] 75/82: fixed

**Comment by japaric** [comment]
thanks for the report. I think this is a rustc / LLVM bug that occurs with the thumbv6m-none-eabi target but not with the other thumb targets. for now you can work around the issue by commenting out the `opt-level` line under `profile.dev` in the `Cargo.toml` file. I believe this issue has already been reported in the rust-lang/rust issue tracker; I'll see if I can locate the ticket.

**Comment by japaric** [comment]
> I'll see if I can locate the ticket.

this is the upstream issue: https://github.com/rust-lang/rust/issues/75045

**Comment by tspiteri** [comment]
I think I managed to work around the rustc/LLVM bug in the fixed crate and plan to release a new version soon which should work around the issue. Can you please check whether it fixes your issue? To test, please use the following in Cargo.toml for the fixed dependency:
```toml
[dependencies]
fixed = { git = "https://gitlab.com/tspiteri/fixed.git", rev = "9be4f1b4" }
```

**Comment by tspiteri** [comment]
For what it's worth, while the tweak I did in the fixed crate was prompted by the rustc/LLVM issue which was reported as [issue 45](https://gitlab.com/tspiteri/fixed/-/issues/45), I wasn't trying to work around the rustc/LLVM issue. I just tried to make two internal functions which were triggering the bug more readable. In the process I converted a couple of expressions of the form `expr + (carry_as_u128 << 64)` which are guaranteed not to overflow into something of the form `expr.wrapping_add(carry_as_u128 << 64)`, and that seems to not trigger the compiler hang.

**Comment by Urhengulas** [comment]
Hi @tspiteri,

Yes, building works for me just fine. I don't even need to specify `fixed` as `1.13`, cargo automatically picks up the new version. But I checked that it is `1.13` in Cargo.lock`.

When pinning `fixed` to `1.12` the build hangs as before.

Thank you for solving this!
